certs/cert.go

```diff
@@ -0,0 +1,193 @@
+	if len(certPath) > 0 && len(keyPath) > 0 && len(domain) > 0 && common.FileExisted(certPath) && common.FileExisted(keyPath) {
+		return VerifyCertificate(certPath, domain)
+	}
+	return false, nil
+}
+
+func CertExists() (bool, bool, bool) {
+	wssUserCertExists, err := VerifyLocalCertificate(config.Parameters.HttpWssCert, config.Parameters.HttpWssKey, config.Parameters.HttpWssDomain)
+	if err != nil {
+		log.Error(err)
+	}
+	httpsUserCertExists, err := VerifyLocalCertificate(config.Parameters.HttpsJsonCert, config.Parameters.HttpsJsonKey, config.Parameters.HttpsJsonDomain)
+	if err != nil {
+		log.Error(err)
+	}
+	prePreparedCert := len(config.Parameters.HttpsJsonDomain) > 0 && len(config.Parameters.HttpWssDomain) > 0
+	defaultDomain, err := node.GetDefaultDomainFromIP(config.Parameters.Hostname, config.Parameters.DefaultTlsDomainTmpl)
+	if err != nil {
+		log.Error(err)
+	}
+	_, _, certFile, keyFile := GetACMEFileNames(defaultDomain, config.Parameters.CertDirectory)
+	defaultCertExists, err := VerifyLocalCertificate(certFile, keyFile, defaultDomain)
+
+	return wssUserCertExists || defaultCertExists, httpsUserCertExists || defaultCertExists, prePreparedCert
+}
+
+func PrepareCert(wssCertExists, httpsCertExists, prePreparedCert bool, wssCertReady, httpsCertReady chan struct{}) {
+	first := true
+	for {
+		changeToDefault(httpsCertExists, wssCertExists, prePreparedCert)
+		cm, err := NewCertManager()
+		if err != nil {
+			log.Errorf("create new certManager err: %v", err)
+			return
+		}
+		if !first {
+			time.Sleep(config.Parameters.CertCheckInterval * time.Second)
+		}
+		first = false
+		if cm.NeedObtain() {
+			cert, err := cm.ObtainNewCert()
+			if err != nil {
+				log.Errorf("apply cert failed: %v", err)
+				continue
+			}
+			err = saveCertToDisk(cert)
+			if err != nil {
+				log.Errorf("save cert failed: %v", err)
+				continue
+			}
+			if !wssCertExists {
+				select {
+				case wssCertReady <- struct{}{}:
+					log.Info("wss cert ready")
+				}
+			}
+			if !httpsCertExists {
+				select {
+				case httpsCertReady <- struct{}{}:
+					log.Info("https cert ready")
+				}
+			}
+		} else if cm.NeedRenew() {
+			renew, err := cm.RenewCert()
+			if err != nil {
+				log.Errorf("renew cert failed: %v", err)
+				continue
+			}
+			err = saveCertToDisk(renew)
+			if err != nil {
+				log.Errorf("save renew cert failed: %v", err)
+				continue
+			}
+		}
+	}
+}
+
+func changeToDefault(https, wss, prePreparedCert bool) {
+	defaultDomain, err := node.GetDefaultDomainFromIP(config.Parameters.Hostname, config.Parameters.DefaultTlsDomainTmpl)
+	if err != nil {
+		log.Error(err)
+	}
+
+	userFile, resourceFile, certFile, keyFile := GetACMEFileNames(defaultDomain, config.Parameters.CertDirectory)
+	config.Parameters.DefaultTlsCert = certFile
+	config.Parameters.DefaultTlsKey = keyFile
+	config.Parameters.ACMEUserFile = userFile
+	config.Parameters.ACMEResourceFile = resourceFile
+	if !https || !prePreparedCert {
```

Comment:
I think the logic is not right. Assume user provides a https certificate but not wss certificate, this if condition should be false, but it will be true currently